client/src/js/components/bhBedSelect/bhBedSelect.js

```diff
@@ -0,0 +1,79 @@
+angular.module('bhima.components')
+  .component('bhBedSelect', {
+    templateUrl : 'js/components/bhBedSelect/bhBedSelect.html',
+    controller  : BedSelectController,
+    transclude  : true,
+    bindings    : {
+      uuid      : '<',
+      onSelectCallback : '&',
+      required : '<?',
+      label    : '@?',
+      autoSelect : '<?',
+      roomUuid : '<?',
+      occupied : '<?',
+    },
+  });
+
+BedSelectController.$inject = ['BedService', 'NotifyService'];
+
+/**
+ * Bed selection component
+ */
+function BedSelectController(Bed, Notify) {
+  const $ctrl = this;
+
+  $ctrl.$onInit = function onInit() {
+    $ctrl.label = $ctrl.label || 'BED.TITLE';
+    $ctrl.disabled = !$ctrl.roomUuid;
+    loadBeds();
+  };
+
+  $ctrl.$onChanges = (changes) => {
+    if (changes.roomUuid && changes.roomUuid.currentValue) {
+      $ctrl.disabled = !changes.roomUuid.currentValue;
+      const roomUuid = changes.roomUuid.currentValue || 'x';
+      loadBeds(roomUuid);
+    }
+
+    if (changes.autoSelect && $ctrl.roomUuid) {
+      $ctrl.disabled = false;
+      loadBeds();
+    }
+  };
+
+  // fires the onSelectCallback bound to the component boundary
+  $ctrl.onSelect = $item => {
+    $ctrl.onSelectCallback({ bed : $item });
+  };
+
+  function loadBeds(roomUuid) {
+    const params = loadParams();
+    params.room_uuid = roomUuid || params.room_uuid;
+    Bed.read(null, params)
+      .then(beds => {
+        $ctrl.beds = beds;
+
+        if ($ctrl.autoSelect && beds.length > 0) {
+          [$ctrl.bed] = beds;
+          $ctrl.bed.hrLabel = $ctrl.bed.ward_name.concat('/', $ctrl.bed.room_label, '/', $ctrl.bed.label);
+          $ctrl.id = $ctrl.bed.id;
+          $ctrl.onSelectCallback({ bed : $ctrl.id });
+        }
+
+        $ctrl.noBedAvailable = !!(beds.length === 0);
+      })
+      .catch(Notify.handleError);
+  }
+
+  function loadParams() {
+    // load only beds for a given room
+    const allBeds = { room_uuid : $ctrl.roomUuid || 'x' };
+    const occupiedBeds = { room_uuid : $ctrl.roomUuid || 'x', is_occupied : 1 };
+    const notOccupiedBeds = { room_uuid : $ctrl.roomUuid || 'x', is_occupied : 0 };
+
+    // eslint-disable-next-line no-nested-ternary
+    return $ctrl.occupied === true ? occupiedBeds
```

Comment:
1. The room_uuid is required, if it missing we must not have beds listed. I set 'x' as default because if it null or undefined the filter on the server side returns all beds.

2. "$ctrl.occupied" is in three states because it works like a filter, if true it will show occupied beds, if false it will show just free beds (not occupied), if undefined it will show all beds